src/key_exchange/tripledh.rs

```diff
@@ -0,0 +1,415 @@
+                server_e_pk: server_e_kp.public().clone(),
+                mac: mac.finalize().into_bytes().to_vec(),
+            },
+        ))
+    }
+
+    fn generate_ke3<KeyFormat: KeyPair<Repr = Key>>(
+        l2_component: Vec<u8>,
+        ke2_message: Self::KE2Message,
+        ke1_state: &Self::KE1State,
+        server_s_pk: KeyFormat::Repr,
+        client_s_sk: KeyFormat::Repr,
+    ) -> Result<(Vec<u8>, Self::KE3Message), ProtocolError> {
+        let (shared_secret, km2, km3) = derive_3dh_keys::<KeyFormat>(
+            TripleDHComponents {
+                pk1: ke2_message.server_e_pk.clone(),
+                sk1: ke1_state.client_e_sk.clone(),
+                pk2: server_s_pk.clone(),
+                sk2: ke1_state.client_e_sk.clone(),
+                pk3: ke2_message.server_e_pk.clone(),
+                sk3: client_s_sk.clone(),
+            },
+            &ke1_state.client_nonce,
+            &ke2_message.server_nonce,
+            KeyFormat::public_from_private(&client_s_sk),
+            server_s_pk,
+        )?;
+
+        let transcript: Vec<u8> = [
+            &ke1_state.hashed_l1[..],
+            &l2_component[..],
+            &ke2_message.server_nonce[..],
+            &ke2_message.server_e_pk[..],
+        ]
+        .concat();
+
+        let mut hasher = Sha256::new();
+        hasher.update(&transcript);
+        let hashed_transcript = hasher.finalize();
+
+        let mut server_mac =
+            Hmac::<Sha256>::new_varkey(&km2).map_err(|_| InternalPakeError::HmacError)?;
+        server_mac.update(&hashed_transcript);
+
+        if ke2_message.mac != server_mac.finalize().into_bytes().to_vec() {
+            return Err(ProtocolError::VerificationError(
+                PakeError::KeyExchangeMacValidationError,
+            ));
+        }
+
+        let mut client_mac =
+            Hmac::<Sha256>::new_varkey(&km3).map_err(|_| InternalPakeError::HmacError)?;
+        client_mac.update(&hashed_transcript);
+
+        Ok((
+            shared_secret.to_vec(),
+            KE3Message {
+                mac: client_mac.finalize().into_bytes().to_vec(),
+            },
+        ))
+    }
+
+    fn finish_ke(
+        ke3_message: Self::KE3Message,
+        ke2_state: &Self::KE2State,
+    ) -> Result<Vec<u8>, ProtocolError> {
+        let mut client_mac =
+            Hmac::<Sha256>::new_varkey(&ke2_state.km3).map_err(|_| InternalPakeError::HmacError)?;
+        client_mac.update(&ke2_state.hashed_transcript);
+
+        if ke3_message.mac != client_mac.finalize().into_bytes().to_vec() {
+            return Err(ProtocolError::VerificationError(
+                PakeError::KeyExchangeMacValidationError,
+            ));
+        }
+
+        Ok(ke2_state.shared_secret.to_vec())
+    }
+
+    fn ke1_state_size() -> usize {
+        KE1_STATE_LEN
+    }
+
+    fn ke2_message_size() -> usize {
+        KE2_MESSAGE_LEN
+    }
+}
+
+#[derive(PartialEq, Eq)]
+pub struct KE1State {
+    client_e_sk: Key,
```

Comment:
https://tools.ietf.org/html/draft-krawczyk-cfrg-opaque-06#section-6
Does that make sense?